### ISSUE TITLE
1000 epochs to train 

### ISSUE THREAD

**Issue by jananir2202** [msahasrabudhe/crosswise_sparse_autoencoder_pytorch#1]
The number of epochs is set as 1000 in the yaml file while the paper quotes 6 epochs. Can you let me know where I am going wrong? It would also be useful if I could know how long it took to train (with how many epochs) and also the number of GPUs used.

**Comment by msahasrabudhe** [comment]
HI jananir2202,

Sorry, I am not sure I understand the problem you're facing. The number of epochs in the configuration file was set to 1000 to see how the model performs. In the paper, the authors use a different dataset and found that 6 epochs were enough. 

The model was trained on only one GTX 1080 GPU, but I do not unfortunately remember the time it took. I am also unable to restart the experiments right now due to lack of time.

**Comment by jananir2202** [comment]
1000 epochs seemed to be quite a huge amount of time. But I didn't notice that the datasets were different. Thanks for the information. I shall look into it.

**Comment by ciaobladoo** [comment]
@jananir2202 Not sure if this is useful for you after such a long time. But i run the code and it converges fairly quickly. To get qualitatively good results, you only need 1~2 epoch it seems, which is approxiamtely 30min/epoch on my GPU. However, I think the current code did not enforce sparsity properly and had to make some changes myself. With the current code, I can get an almost perfect reconstruction, but the density of detection map is approximately 0.8, which is very dense, so all the reconstruction comes from the forground with the background ignored. This is essentially just an normal autoencoder that does not seperate nucleus and background. The changes I had to make are the following: 1, you will need another BatchNorm2d for the DetectionBranch in model.py, applied after self.conv2. Specifically, I set "affine=False" to this BatchNorm; 2. tau should be calculated across the minibatch instead of being an average of all the images (line 96-100 in model.py); 3, you need to change the parameter "p" in the config accordingly. I made these changes according to the code from the original authors (I downloaded a long time ago, I don't think I can find it now). And with that I am able to get results similiar to what's shown in the repo with in 1 epoch. I think @msahasrabudhe wrote an excellent code, just want to point out this minor issue.

P.S. I just realize after the modification, if I run it long,er like 3 epochs, the reconstruction becomes better, but the detection map becomes dense again. I think it's because of the moving average.

**Comment by jananir2202** [comment]
Hi,

Thank you very much for remembering all the way and replying back. Indeed a
great gesture. I shall definitely look into it again.

Thanks and Regards
Janani Ramaswamy

**Comment by hiraksarkar** [comment]
@ciaobladoo Hi do you have the forked code with the fixes by any chance?

**Comment by ciaobladoo** [comment]
> @ciaobladoo Hi do you have the forked code with the fixes by any chance?

Sorry man. It's too old. I ended up not using this code. And apparently I have deleted my fix of it. I am happy to give some alternative suggestions depending on what you want to do. You can email me at cfeng07@gmail.com. Blessings~
